pymc4/model.py

```diff
@@ -0,0 +1,126 @@
+        return self
+
+    def __exit__(self, typ, value, traceback):
+        type(self).get_contexts().pop()
+
+
+def withparent(meth):
+    """Helper wrapper that passes calls to parent's instance"""
+    def wrapped(self, *args, **kwargs):
+        res = meth(self, *args, **kwargs)
+        if getattr(self, 'parent', None) is not None:
+            getattr(self.parent, meth.__name__)(*args, **kwargs)
+        return res
+    # Unfortunately functools wrapper fails
+    # when decorating built-in methods so we
+    # need to fix that improper behaviour
+    wrapped.__name__ = meth.__name__
+    return wrapped
+
+
+class treedict(dict):
+    """A dict that passes mutable extending operations used in Model
+    to parent dict instance.
+    Extending treedict you will also extend its parent
+    """
+    def __init__(self, iterable=(), parent=None, **kwargs):
+        super(treedict, self).__init__(iterable, **kwargs)
+        assert isinstance(parent, dict) or parent is None
+        self.parent = parent
+        if self.parent is not None:
+            self.parent.update(self)
+    # typechecking here works bad
+    __setitem__ = withparent(dict.__setitem__)
+    update = withparent(dict.update)
+
+    def tree_contains(self, item):
+        # needed for `add_random_variable` method
+        if isinstance(self.parent, treedict):
+            return (dict.__contains__(self, item) or
+                    self.parent.tree_contains(item))
+        elif isinstance(self.parent, dict):
+            return (dict.__contains__(self, item) or
+                    self.parent.__contains__(item))
+        else:
+            return dict.__contains__(self, item)
+
+
+class Model(Context):
+    def __new__(cls, *args, **kwargs):
+        instance = super(Model, cls).__new__(cls)
+        if kwargs.get('model') is not None:
+            instance.parent = kwargs.get('model')
+        elif cls.get_contexts():
+            instance.parent = cls.get_contexts()[-1]
+        else:
+            instance.parent = None
+        return instance
+
+    def __init__(self, name="", model=None, ):
+        self.name = name
+        if self.parent is not None:
+            self.named_vars = treedict(parent=self.parent.named_vars)
+        else:
+            self.named_vars = treedict()
+
+    @property
+    def model(self):
+        return self
+
+    @property
+    def decription(self):
+        return
+
+    @classmethod
+    def get_contexts(cls):
+        # no race-condition here, cls.contexts is a thread-local object
+        # be sure not to override contexts in a subclass however!
+        if not hasattr(cls.contexts, 'stack'):
+            cls.contexts.stack = []
+        return cls.contexts.stack
+
+    @classmethod
+    def get_context(cls):
+        """Return the deepest context on the stack."""
+        try:
+            return cls.get_contexts()[-1]
+        except IndexError:
+            raise TypeError("No context on context stack")
```

Comment:
This could be more informative: "This must be called from inside a `Model` context manager"